src/db_cache/mod.rs

```diff
@@ -0,0 +1,77 @@
+use std::sync::Arc;
+
+use async_trait::async_trait;
+
+use crate::{
+    block::Block, db_state::SsTableId, filter::BloomFilter, flatbuffer_types::SsTableIndexOwned,
+};
+
+/// The default max capacity for the cache. (64MB)
+pub const DEFAULT_MAX_CAPACITY: u64 = 64 * 1024 * 1024;
+/// The default cached block size for the cache. (32 bytes)
+pub const DEFAULT_CACHED_BLOCK_SIZE: u32 = 32;
+
+#[cfg(feature = "foyer")]
+pub mod foyer;
+#[cfg(feature = "moka")]
+pub mod moka;
+
+/// A trait for in-memory caches.
+///
+/// This trait defines the interface for an in-memory cache,
+/// which is used to store and retrieve cached blocks associated with SSTable IDs.
+#[async_trait]
+pub trait DbCache: Send + Sync {
+    async fn get(&self, key: (SsTableId, u64)) -> Option<CachedEntry>;
+    async fn insert(&self, key: (SsTableId, u64), value: CachedEntry);
+    #[allow(dead_code)]
+    async fn remove(&self, key: (SsTableId, u64));
+    #[allow(dead_code)]
+    fn entry_count(&self) -> u64;
+}
+
+/// A cached entry stored in the cache.
+#[derive(Clone, Default)]
+pub struct CachedEntry {
+    block: Option<Arc<Block>>,
+    sst_index: Option<Arc<SsTableIndexOwned>>,
+    bloom_filter: Option<Arc<BloomFilter>>,
+}
+
+impl CachedEntry {
+    /// Create a new `CachedEntry` with the given block.
+    pub(crate) fn with_block(self, block: Arc<Block>) -> Self {
```

Comment:
It seems better to just have methods like `CachedEntry::bloom_filter, CachedEntry::block, CachedEntry::index` as it's not valid for an entry to contain all None fields or more than one field with some value.

Alternatively, we can keep the old `CachedBlock` enum and just make it a field in `CachedEntry` so that we don't have to expose all the types used in the enum publicly. It's really an enum that we want anyway.

Reply:
changed.